File: cses/sliding_window/Sliding Window Inversions.py
from sortedcontainers import SortedList
n,k=[int(x) for x in input().split()]
arr=[int(x) for x in input().split()]
class Map:
    def __init__(self):
        self.l = SortedList()

    def insert(self, val):
        lesser = self.l.bisect_left(val)  # Elements strictly less than val
        greater = len(self.l) - self.l.bisect_right(val)  # Elements strictly greater than val
        self.l.add(val)
        return [lesser, greater]

    def remove(self, val):
        if val not in self.l:
            return [-1, -1]  # Or raise an error if desired
        lesser = self.l.bisect_left(val)
        greater = len(self.l) - self.l.bisect_right(val)
        self.l.remove(val)
        return [lesser, greater]
obj=Map()
count=0
for i in range(k):
    count+=obj.insert(arr[i])[1]
print(count,end=" ")
for i in range(k,n):
    count-=obj.remove(arr[i-k])[0]
    count+=obj.insert(arr[i])[1]
    print(count,end=" ")

File: cses/sliding_window/sliding_window_inversions.cpp
#include <iostream>
#include <set>
#include <vector>

using namespace std;

class Map {
    multiset<int> ms;

public:
    pair<int, int> insert(int val) {
        int lesser = distance(ms.begin(), ms.lower_bound(val));
        int greater = distance(ms.upper_bound(val), ms.end());
        ms.insert(val);
        return {lesser, greater};
    }

    pair<int, int> remove(int val) {
        auto it = ms.find(val);
        if (it == ms.end()) return {-1, -1};
        int lesser = distance(ms.begin(), ms.lower_bound(val));
        int greater = distance(ms.upper_bound(val), ms.end());
        ms.erase(it);  // Only removes one occurrence
        return {lesser, greater};
    }
};

int main() {
    int n, k;
    cin >> n >> k;
    vector<int> arr(n);
    for (int &x : arr) cin >> x;

    Map obj;
    int count = 0;

    for (int i = 0; i < k; ++i) {
        count += obj.insert(arr[i]).second;
    }
    cout << count << " ";

    for (int i = k; i < n; ++i) {
        count -= obj.remove(arr[i - k]).first;
        count += obj.insert(arr[i]).second;
        cout << count << " ";
    }

    return 0;
}
